networkx/algorithms/summarization.py

```diff
@@ -210,3 +210,456 @@ def dedensify(G, threshold, prefix=None, copy=True):
+                            has_edges
+                            and output.get_edge_data(*summary_edge)
+                            == superedge_attributes
+                        ):
+                            continue
+
+                    output.add_edge(*summary_edge, **superedge_attributes)
+
+    nx.set_node_attributes(output, supernode_lookup, supernode_attribute)
+    return output
+
+
+def snap_build_data_structures(G, node_attributes, edge_attributes):
+    """Returns data structures needed to build the SNAP summary graph
+
+    Creates the data structures and splits the node groups  until all are groups
+    are homogenous in terms of node attribute values and edge types to
+    build the output graph for the SNAP summary graph.
+
+    Parameters
+    ----------
+    G: graph
+        Networkx Graph to be summarized
+    node_attributes: iterable
+        The node attributes that can be found in nodes in the graph, and are to
+        be recognized in the summarization
+    edge_attributes: iterable
+        The edge attributes that can be found in edges in the graph, and
+        are to be recognized in the summarization
+
+    Returns
+    -------
+    tuple of data structure representations: (list, list, dict, dict)
+
+    Notes
+    -----
+    Finds the AR-compatible grouping of graph `G` and returns data structures
+    to efficiently process those groupings.
+
+    An Attribute-edge compatible (AR-compatible) grouping is where
+    all nodes in each group have the same exact node attribute values and
+    have the same exact edge types to nodes in the same exact
+    groups.
+    """
+    edge_types = {
+        edge: tuple(attrs.get(attr) for attr in edge_attributes)
+        for edge, attrs in G.edges.items()
+    }
+    if not G.is_directed():
+        if G.is_multigraph():
+            edge_types.update(
+                ((e[1], e[0], e[2]), edge_types[e]) for e in list(edge_types)
+            )
+        else:
+            edge_types.update(
+                ((v, u), edge_types[(u, v)]) for (u, v) in list(edge_types)
+            )
+
+    group_lookup = {
+        node: tuple(attrs[attr] for attr in node_attributes)
+        for node, attrs in G.nodes.items()
+    }
+    groups = defaultdict(set)
+    for node, node_type in group_lookup.items():
+        groups[node_type].add(node)
+
+    neighbor_groups = snap_calculate_neighbor_groups(
+        G,
+        groups,
+        edge_types,
+        group_lookup,
+    )
+    eligible_groups = [
+        g_id for g_id in groups if snap_is_eligible(groups, neighbor_groups, g_id)
+    ]
+    while eligible_groups:
+        group_id = eligible_groups.pop()
+        groups = snap_split(
+            G, groups, neighbor_groups, edge_types, group_lookup, group_id
+        )
+        neighbor_groups = snap_calculate_neighbor_groups(
+            G,
+            groups,
+            edge_types,
+            group_lookup,
+        )
+        eligible_groups = [
```

Comment:
Do we need to check all the `snap_is_eligible` when we are just going to use the first one (with `pop`)?
Perhaps we could reduce the computation considerably here...